File: nodejs/tests/test_three_routers.js
const chai = require('chai')
const assert = chai.assert
const { describe, it } = require('mocha')
const { Packet } = require('../lib/aln/packet')
const Router = require('../lib/aln/router')

// LocalChannel implementation - simulates a bidirectional communication channel
// Each channel has a twin, and sends packets to its twin with a small async delay
class LocalChannel {
  constructor (twin) {
    this.twin = twin || new LocalChannel(this)
    this.onPacket = (packet) => {}
    this.onClose = null
  }

  send (packet) {
    // Simulate async packet delivery
    setTimeout(() => {
      if (this.twin && this.twin.onPacket) {
        this.twin.onPacket(packet)
      }
    }, 5)
  }

  close () {
    if (this.onClose) {
      this.onClose()
    }
  }
}

// Helper to wait for an event with timeout
function waitForEvent (checkFn, timeoutMs = 5000, checkIntervalMs = 50) {
  return new Promise((resolve, reject) => {
    const startTime = Date.now()
    const interval = setInterval(() => {
      if (checkFn()) {
        clearInterval(interval)
        resolve(true)
      } else if (Date.now() - startTime > timeoutMs) {
        clearInterval(interval)
        reject(new Error('Timeout waiting for event'))
      }
    }, checkIntervalMs)
  })
}

// Helper to wait for a specified time
function wait (ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

describe('# Three Router Topology Tests', function () {
  this.timeout(15000) // Increase timeout for network discovery

  it('# ping-pong across three routers (multi-hop)', async function () {
    // Create three routers in a chain topology:
    // Router1 <---> Router2 <---> Router3
    //
    // Router1 has the "ping" service
    // Router3 will discover and send a ping request
    // Router2 acts as a relay between them

    // Setup Router 1 and Router 2 connection
    const ch1a = new LocalChannel()
    const ch2a = ch1a.twin

    // Track ping handler execution
    let receivedPing = false
    let pingResponseSent = false
    let receivedPingData = null

    // Create Router 1 with ping service
    const router1 = new Router('aln-node-1')
    router1.registerService('ping', (packet) => {
      console.log('Router 1: Received ping request')
      receivedPing = true
      receivedPingData = packet.data

      // Send pong response back to the source
      const pongPacket = new Packet()
      pongPacket.dst = packet.src
      pongPacket.ctx = packet.ctx
      pongPacket.data = 'pong'
      router1.send(pongPacket)
      pingResponseSent = true
      console.log('Router 1: Sent pong response')
    })
    router1.addChannel(ch1a)

    // Create Router 2 (relay)
    const router2 = new Router('aln-node-2')
    router2.addChannel(ch2a)

    // Setup Router 2 and Router 3 connection
    const ch1b = new LocalChannel()
    const ch2b = ch1b.twin

    // Track pong response reception
    let receivedPong = false
    let pongData = null

    // Create Router 3 (client)
    const router3 = new Router('aln-node-3')

    // Register context handler for pong response
    const ctxID = router3.registerContextHandler((packet) => {
      console.log('Router 3: Received pong response')
      pongData = packet.data
      receivedPong = true
    })

    // Track service discovery
    let pingServiceDiscovered = false
    let pingServiceAddress = null

    router3.setOnServiceCapacityChanged((service, capacity, address) => {
      console.log(`Router 3: Service discovered - ${service} at ${address} (capacity: ${capacity})`)
      if (service === 'ping' && capacity > 0) {
        pingServiceAddress = address
        pingServiceDiscovered = true
      }
    })

    // Connect Router 3 and Router 2
    router3.addChannel(ch2b)
    router2.addChannel(ch1b)

    console.log('All routers connected, waiting for network discovery...')

    // Wait for initial network discovery
    await wait(500)

    // Manually trigger network state sharing to ensure discovery
    router1.shareNetState()
    router2.shareNetState()
    router3.shareNetState()

    // Wait for route propagation through all hops
    await wait(1000)

    // Trigger again to ensure full propagation
    router1.shareNetState()
    router2.shareNetState()
    router3.shareNetState()

    await wait(500)

    // Wait for service discovery with timeout
    console.log('Waiting for ping service discovery...')
    await waitForEvent(() => pingServiceDiscovered, 10000)
    assert.isTrue(pingServiceDiscovered, 'Ping service should be discovered')
    assert.equal(pingServiceAddress, 'aln-node-1', 'Ping service should be at aln-node-1')

    console.log('Ping service discovered! Sending ping request...')

    // Send ping request from Router 3
    const pingPacket = new Packet()
    pingPacket.srv = 'ping'
    pingPacket.ctx = ctxID
    pingPacket.data = 'hello world'
    router3.send(pingPacket)

    // Wait for ping to be received by Router 1
    console.log('Waiting for Router 1 to receive ping...')
    await waitForEvent(() => receivedPing, 5000)
    assert.isTrue(receivedPing, 'Router 1 should receive ping')
    assert.equal(receivedPingData, 'hello world', 'Ping data should match')

    // Wait for Router 1 to send pong response
    console.log('Waiting for Router 1 to send pong...')
    await waitForEvent(() => pingResponseSent, 5000)
    assert.isTrue(pingResponseSent, 'Router 1 should send pong response')

    // Wait for Router 3 to receive pong response
    console.log('Waiting for Router 3 to receive pong...')
    await waitForEvent(() => receivedPong, 5000)
    assert.isTrue(receivedPong, 'Router 3 should receive pong')
    assert.equal(pongData, 'pong', 'Pong data should be "pong"')

    console.log('✅ Three-router ping-pong test passed!')
  })

  it('# service multicast across three routers', async function () {
    // Create three routers where Router1 and Router3 both have the same service
    // Router2 in the middle should route to both

    // Setup Router 1
    const ch1a = new LocalChannel()
    const ch2a = ch1a.twin

    let router1Received = false
    const router1 = new Router('aln-node-1')
    router1.registerService('echo', (packet) => {
      console.log('Router 1: Received echo request')
      router1Received = true
    })
    router1.addChannel(ch1a)

    // Setup Router 2 (relay)
    const router2 = new Router('aln-node-2')
    router2.addChannel(ch2a)

    // Setup Router 3
    const ch1b = new LocalChannel()
    const ch2b = ch1b.twin

    let router3Received = false
    const router3 = new Router('aln-node-3')
    router3.registerService('echo', (packet) => {
      console.log('Router 3: Received echo request')
      router3Received = true
    })
    router3.addChannel(ch2b)
    router2.addChannel(ch1b)

    console.log('All routers connected, waiting for network discovery...')

    // Wait for network discovery
    await wait(500)
    router1.shareNetState()
    router2.shareNetState()
    router3.shareNetState()
    await wait(1000)
    router1.shareNetState()
    router2.shareNetState()
    router3.shareNetState()
    await wait(500)

    console.log('Sending multicast echo request from Router 2...')

    // Send service multicast from Router 2 (should reach both Router 1 and Router 3)
    const echoPacket = new Packet()
    echoPacket.srv = 'echo'
    echoPacket.data = 'broadcast message'
    router2.send(echoPacket)

    // Wait for both routers to receive the message
    console.log('Waiting for multicast delivery...')
    await waitForEvent(() => router1Received && router3Received, 5000)

    assert.isTrue(router1Received, 'Router 1 should receive echo')
    assert.isTrue(router3Received, 'Router 3 should receive echo')

    console.log('✅ Service multicast test passed!')
  })
})

File: python/tests/test_three_routers.py
import pytest
import os
import selectors
import sys
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..' , 'src')))
from alnmeshpy import LocalChannel, Router, Packet
from threading import Event

# Fixture for setting up routers and channels
@pytest.fixture
def three_routers_setup():
    # Setup for Router 1 and its channels
    r1a, w1a = os.pipe()
    r2a, w2a = os.pipe()
    ch1a = LocalChannel(r1a, w2a)
    ch2a = LocalChannel(r2a, w1a)

    sel = selectors.DefaultSelector()

    # Configure the ping handler on node 1
    received_ping = Event()
    ping_response_sent = Event()

    def ping_handler(packet):
        received_ping.set()
        router1.send(Packet(
            destAddr=packet.srcAddr,
            contextID=packet.contextID,
            data="pong"
        ))
        ping_response_sent.set()

    router1 = Router(sel, "aln-node-1")
    router1.start()
    router1.register_service("ping", ping_handler)
    router1.add_channel(ch1a)

    # Setup for Router 2
    router2 = Router(sel, "aln-node-2")
    router2.start()
    router2.add_channel(ch2a)

    # Setup for Router 3 and its channels
    r1b, w1b = os.pipe()
    r2b, w2b = os.pipe()
    ch1b = LocalChannel(r1b, w2b)
    ch2b = LocalChannel(r2b, w1b)

    received_pong = Event()
    pong_data = {"data": None}  # Use dict so changes are reflected

    def on_packet(packet):
        pong_data["data"] = packet.data
        received_pong.set()

    router3 = Router(sel, "aln-node-3")
    router3.start()
    ctxID = router3.register_context_handler(on_packet)

    # Flag to control when to send ping
    ping_service_available = Event()
    ping_service_address = {"addr": None}

    def on_service_discovery(service, load, address):
        if service == "ping":
            ping_service_address["addr"] = address
            ping_service_available.set()

    router3.set_on_service_load_changed_handler(on_service_discovery)
    router3.add_channel(ch2b)
    router2.add_channel(ch1b)

    yield router1, router2, router3, sel, received_ping, ping_response_sent, received_pong, pong_data, ctxID, ping_service_available

    # Teardown
    router1.stop()
    router2.stop()
    router3.stop()
    ch1a.close()
    ch2a.close()
    ch1b.close()
    ch2b.close()
    sel.close()
    # File descriptors are already closed by LocalChannel.close() calls above


def test_three_routers_ping_pong(three_routers_setup):
    router1, router2, router3, sel, received_ping, ping_response_sent, received_pong, pong_data, ctxID, ping_service_available = three_routers_setup

    # Give some time for network discovery to complete
    import time
    time.sleep(0.5)

    # Manually trigger network state sharing to ensure service discovery
    router1.share_net_state()
    router2.share_net_state()
    router3.share_net_state()

    # Give more time for route discovery to propagate through all hops
    time.sleep(2.0)

    # Trigger again to ensure route propagation
    router1.share_net_state()
    router2.share_net_state()
    router3.share_net_state()

    time.sleep(1.0)

    # Wait for service discovery to complete
    assert ping_service_available.wait(timeout=10), "Ping service was not discovered"

    # Now send the ping manually from the test
    text = "hello world".encode(encoding="utf-8")
    router3.send(Packet(service="ping", contextID=ctxID, data=text))

    # Wait for the ping to be received by router1
    assert received_ping.wait(timeout=10), "Router 1 did not receive ping"
    # Wait for router1 to send the pong response
    assert ping_response_sent.wait(timeout=10), "Router 1 did not send pong response"
    # Wait for router3 to receive the pong response
    assert received_pong.wait(timeout=10), "Router 3 did not receive pong"

    assert pong_data["data"] is not None, "No pong data received"
    assert pong_data["data"].decode(encoding="utf-8") == "pong", f"Expected 'pong', got {pong_data['data'].decode(encoding='utf-8')}"
